gapis/resolve/dependencygraph2/README.md

```diff
@@ -0,0 +1,165 @@
+depend on both node 11 and 22. As it writes to slot2, at the end of the command
+mutation, the memory accesses tracker is updated to:
+
+```
+slot1: 11
+slot2: 33
+```
+
+This makes sure that the next command to read from slot2 will depend on 33.
+
+Now, a single command may have multiple read/write to the same slot: we should
+not update the memory access tracker as soon as there is a write, otherwise a
+subsequent read by the same command would create a circular dependency to the
+node itself. To avoid this, all read/writes of a single command are temporarily
+stored in a list of pending accesses, which are flushed when the command ends.
+
+### In practice
+
+In practice, in `BuildDependencyGraph()`, a `dependencyGraphBuilder` is created
+and passed as the `api.StateWatcher` to each command's `Mutate()`. The
+`api.StateWatcher` interface lists callbacks that are invoked during mutation,
+including when a command or subcommand starts or ends, and whenever it does an
+API state access (`OnReadFrag()`/`OnWriteFrag()`) or a memory access
+(`OnReadSlice()`/`OnWriteSlice()` and `OnReadObs()`/`OnWriteObs()`).
+
+The `dependencyGraphBuilder` builder object contains itself some dedicated
+sub-watchers like a `FragWatcher` and a `MemWatcher`, which focus on API state
+and memory, respectively. These watcher accumulate pending accesses of a command
+and its potential subcommands (so there can be several nodes pending), and
+dependencies are established when a top-level command ends: `OnEndCmd()` calls
+`AddDependencies()` that calls `AddNodeDependencies()` on each pending node.
+
+The tracking of API state fragment and memory accesses is done in an
+API-agnostic way, there is no direct correlation with e.g. Vulkan state or
+objects. This tracking is also more complex than the simplified memory slots
+used to illustrate the approach. basically, a memory access is defined by an
+`interval.U64Span` within a given `memory.PoolID`, and an API state fragment
+access is defined by some `api.RefID` and `api.Fragment`. In both cases, the
+high-level approach presented above applies.
+
+### Forward dependencies
+
+Forward dependencies are a special kind of dependency that applies to API
+commands that come in pairs. For instance, `vkBeginCommandBuffer()` and
+`vkEndCommandBuffer()` come in pairs: if you have one, you should have the
+other.
+
+In practice, alongside the `FragWatcher` and the `MemWatcher`, the dependency
+graph buidler also has a `ForwardWatcher` which handles forward dependencies via
+three callbacks: `OpenForwardDependency()`, `CloseForwardDependency()` and
+`DropForwardDependency()`.
+
+Most forward dependencies are only concerned with open and close operations. For
+instance, the mutation of `vkBeginCommandBuffer()` triggers
+`OpenForwardDependency()` which creates an open forward dependency, and
+`vkBeginCommandBuffer()` triggers `CloseForwardDependency()` which closes it,
```

Comment:
"`vkBeginCommandBuffer()` triggers `CloseForwardDependency()` which closes it,"

You mean:

"`vk**End**CommandBuffer()` triggers `CloseForwardDependency()` which closes it,"

?

Reply:
Good catch, thanks! corrected.